File: frontend/src/services/mockInterviewService.ts
import api from "./api";

/* ==========================================================
   Interfaces
========================================================== */

export interface InterviewQuestion {
  id: number;
  subject: string;
  topic: string;
  difficulty: string;
  question: string;
}

export interface AnswerSubmission {
  question_id: number;
  answer: string;
}

export interface EvaluationCriterion {
  name: string;
  score: number;
  feedback: string;
}

export interface AnswerEvaluation {
  overall_score: number;
  strengths: string[];
  improvements: string[];
  ai_feedback: string;
  criteria: EvaluationCriterion[];
}

export interface InterviewSession {
  session_id: number;
  subject: string;
  total_questions: number;
  completed_questions: number;
  average_score: number;
}

export interface InterviewHistory {
  session_id: number;
  subject: string;
  date: string;
  score: number;
}

export interface InterviewDashboard {
  current_session: InterviewSession;
  history: InterviewHistory[];
}

/* ==========================================================
   Service
========================================================== */

class MockInterviewService {
  async getDashboard(
    userId: number
  ): Promise<InterviewDashboard> {
    const response = await api.get(
      "/mock-interview/dashboard",
      {
        params: {
          user_id: userId,
        },
      }
    );

    return response.data;
  }

  async getQuestions(
    subject: string,
    difficulty: string = "Medium"
  ): Promise<InterviewQuestion[]> {
    const response = await api.get(
      "/mock-interview/questions",
      {
        params: {
          subject,
          difficulty,
        },
      }
    );

    return response.data;
  }

  async evaluateAnswer(
    submission: AnswerSubmission
  ): Promise<AnswerEvaluation> {
    const response = await api.post(
      "/mock-interview/evaluate",
      submission
    );

    return response.data;
  }

  async getCurrentSession(
    userId: number
  ): Promise<InterviewSession> {
    const response = await api.get(
      "/mock-interview/current-session",
      {
        params: {
          user_id: userId,
        },
      }
    );

    return response.data;
  }

  async getHistory(
    userId: number
  ): Promise<InterviewHistory[]> {
    const response = await api.get(
      "/mock-interview/history",
      {
        params: {
          user_id: userId,
        },
      }
    );

    return response.data;
  }
}

export default new MockInterviewService();

File: backend/app/services/mock_interview_service.py
from datetime import date
from sqlalchemy.orm import Session


class MockInterviewService:
    """
    Service layer for the AI Mock Interview module.

    This implementation returns structured mock data.
    Replace the mock logic with database queries and
    LLM-powered evaluation in production.
    """

    @staticmethod
    def get_questions(
        db: Session,
        subject: str,
        difficulty: str,
    ):
        return [
            {
                "id": 1,
                "subject": subject,
                "topic": "Indian Constitution",
                "difficulty": difficulty,
                "question": (
                    "Explain the Fundamental Rights provided under "
                    "the Constitution of India."
                ),
            },
            {
                "id": 2,
                "subject": subject,
                "topic": "Directive Principles",
                "difficulty": difficulty,
                "question": (
                    "Differentiate between Fundamental Rights and "
                    "Directive Principles of State Policy."
                ),
            },
            {
                "id": 3,
                "subject": subject,
                "topic": "Parliament",
                "difficulty": difficulty,
                "question": (
                    "Explain the powers and functions of the Lok Sabha."
                ),
            },
        ]

    @staticmethod
    def evaluate_answer(
        db: Session,
        question_id: int,
        answer: str,
    ):
        length = len(answer.strip())

        overall_score = min(
            100,
            max(40, (length // 5)),
        )

        criteria = [
            {
                "name": "Concept Accuracy",
                "score": 8.5,
                "feedback": (
                    "The answer covers the major concepts correctly."
                ),
            },
            {
                "name": "Completeness",
                "score": 8.0,
                "feedback": (
                    "Include additional examples for a stronger answer."
                ),
            },
            {
                "name": "Structure",
                "score": 9.0,
                "feedback": (
                    "Well-organized with a logical flow."
                ),
            },
            {
                "name": "Language",
                "score": 8.0,
                "feedback": (
                    "Language is clear and easy to understand."
                ),
            },
        ]

        return {
            "overall_score": overall_score,
            "strengths": [
                "Good conceptual understanding.",
                "Logical answer structure.",
                "Relevant terminology used.",
            ],
            "improvements": [
                "Add real-world examples.",
                "Improve conclusion.",
                "Mention constitutional articles where applicable.",
            ],
            "ai_feedback": (
                "Your answer demonstrates a good understanding of "
                "the topic. To achieve a higher score, provide "
                "additional examples, relevant constitutional "
                "provisions, and a concise concluding summary."
            ),
            "criteria": criteria,
        }

    @staticmethod
    def get_current_session(
        db: Session,
        user_id: int,
    ):
        return {
            "session_id": 1001,
            "subject": "Polity",
            "total_questions": 10,
            "completed_questions": 4,
            "average_score": 82.5,
        }

    @staticmethod
    def get_history(
        db: Session,
        user_id: int,
    ):
        return [
            {
                "session_id": 996,
                "subject": "History",
                "date": str(date.today()),
                "score": 78.0,
            },
            {
                "session_id": 997,
                "subject": "Geography",
                "date": str(date.today()),
                "score": 84.5,
            },
            {
                "session_id": 998,
                "subject": "Polity",
                "date": str(date.today()),
                "score": 88.0,
            },
        ]

    @staticmethod
    def get_dashboard(
        db: Session,
        user_id: int,
    ):
        return {
            "current_session": (
                MockInterviewService.get_current_session(
                    db,
                    user_id,
                )
            ),
            "history": (
                MockInterviewService.get_history(
                    db,
                    user_id,
                )
            ),
        }